viamonvif/device/device_test.go

```diff
@@ -0,0 +1,79 @@
+package device
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"strings"
+	"testing"
+	"time"
+
+	"go.viam.com/rdk/logging"
+	"go.viam.com/test"
+)
+
+func TestSendSoapNoHang(t *testing.T) {
+	logger := logging.NewTestLogger(t)
+
+	t.Run("context cancellation works", func(t *testing.T) {
+		// Channel to coordinate server shutdown
+		done := make(chan struct{})
+		// A server that will hang forever
+		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// For GetCapabilities request during initialization, we must return a valid response
+			if strings.Contains(r.Header.Get("Content-Type"), "soap") &&
+				strings.Contains(readBody(r), "GetCapabilities") {
+				w.Write([]byte(`<?xml version="1.0" encoding="UTF-8"?>
+					<SOAP-ENV:Envelope xmlns:SOAP-ENV="http://www.w3.org/2003/05/soap-envelope">
+						<SOAP-ENV:Body>
+							<GetCapabilitiesResponse>
+								<Capabilities>
+									<Media>
+										<XAddr>http://example.com/onvif/media</XAddr>
+									</Media>
+									<PTZ>
+										<XAddr>http://example.com/onvif/ptz</XAddr>
+									</PTZ>
+								</Capabilities>
+							</GetCapabilitiesResponse>
+						</SOAP-ENV:Body>
+					</SOAP-ENV:Envelope>`))
+				return
+			}
+			<-done
+		}))
+		defer func() {
+			close(done)
+			server.Close()
+		}()
+
+		serverURL, err := url.Parse(server.URL)
+		test.That(t, err, test.ShouldBeNil)
+
+		ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+		defer cancel()
+
+		// Create device with the context
+		dev, err := NewDevice(Params{
+			Xaddr:      serverURL,
+			HTTPClient: &http.Client{},
+			Context:    ctx,
+		}, logger)
+		test.That(t, err, test.ShouldBeNil)
+
+		_, err = dev.sendSoap(server.URL, "test message")
+		test.That(t, err, test.ShouldNotBeNil)
+		test.That(t, err.Error(), test.ShouldContainSubstring, "context deadline exceeded")
+	})
+}
+
+// Helper function to read request body.
+func readBody(r *http.Request) string {
+	body, _ := io.ReadAll(r.Body)
+	r.Body.Close()
+	r.Body = io.NopCloser(bytes.NewBuffer(body))
```

Comment:
Why this noopcloser?

Reply:
Also there is a string buffer if you want to return a string

Reply:
> Why this noopcloser?

It allows for the body to be read multiple times. By default, the Body of an http.Request is a `ReadCloser`, which means it can only be read once. I think we can remove it though since we only read it once in the end.

Reply:
> Also there is a string buffer if you want to return a string
> we can remove it though since we only read it once in the end.

Done and done